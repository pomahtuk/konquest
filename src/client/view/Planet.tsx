import React, { ReactElement } from "react";
import { shallowEqual, useSelector, useDispatch } from "react-redux";
import { Tooltip } from "react-tippy";

import { setDestinationPlanet, setOriginPlanet } from "../actions/game.actions";

import Planet from "../../logic/Planet";

import "react-tippy/dist/tippy.css";
import planetSelectorFunction, { PlanetStoreSlice } from "../selectors/planet.selector";
import RandomImage from "./RandomImage";

export interface PlanetProps {
  planet: Planet;
}

const PlanetElement = ({ planet }: PlanetProps): ReactElement => {
  const { originPlanet, activePlayer, destinationPlanet, currentShipsModifier }: PlanetStoreSlice = useSelector(planetSelectorFunction, shallowEqual);
  const dispatch = useDispatch();

  if (!planet) {
    return <span>Â·</span>;
  }

  const onPlanetSelect = (): void => {
    if (!originPlanet && (!planet.owner || planet.owner.id !== activePlayer.id)) {
      // origin should belong to player
      return;
    } else if (!originPlanet && planet.owner && planet.owner.id === activePlayer.id) {
      dispatch(setOriginPlanet(planet));
    } else if (originPlanet && !destinationPlanet) {
      dispatch(setDestinationPlanet(planet));
    } else if (planet.owner && planet.owner.id === activePlayer.id) {
      dispatch(setOriginPlanet(planet));
      dispatch(setDestinationPlanet(undefined));
    }
  };

  return (
    <Tooltip
      position="top"
      animation="fade"
      arrow="true"
      animateFill="false"
      html={
        <div>
          <div>Owner: {planet.owner ? planet.owner.screenName : "none"}</div>
          <div>Production: {planet.production}</div>
          <div>Kill percent: {planet.killPercent}</div>
          <div>Ships: {planet.ships - (currentShipsModifier[planet.name] || 0)}</div>
        </div>
      }
      style={{
        position: "absolute",
        top: planet.coordinates.x * 60 - 10,
        left: planet.coordinates.y * 60 - 10,
        width: 60,
        height: 60
      }}
    >
      <div
        style={{
          cursor: "pointer",
          position: "relative",
          top: 10,
          left: 10,
          width: 40,
          height: 40
        }}
        onClick={onPlanetSelect}
      >
        <RandomImage imageKey={planet.name}></RandomImage>
        <span
          style={{
            position: "absolute",
            bottom: 0,
            right: 0,
            zIndex: 999
          }}
        >
          {planet.name}
        </span>
        <br />
      </div>
    </Tooltip>
  );
};

export default PlanetElement;
