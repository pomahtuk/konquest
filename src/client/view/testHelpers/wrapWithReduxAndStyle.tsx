import React, { ReactElement } from "react";
import { Provider } from "react-redux";
import storeCreator from "../../stores/game.store";
import ThemeProvider from "../themes/ThemeProvider";
import baseTheme from "../themes/base.theme";
const store = storeCreator();

export const CurrentStore = store;

const wrapWithReduxAndStyle = (element: ReactElement): ReactElement => {
  return (
    <Provider store={store}>
      <ThemeProvider theme={baseTheme}>{element}</ThemeProvider>
    </Provider>
  );
};

export default wrapWithReduxAndStyle;
