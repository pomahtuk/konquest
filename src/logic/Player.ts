export default class Player {
  public id: number;
  public screenName: string;

  public constructor(name: string) {
    // TODO: player actually have to be stored somewhere
    this.id = Math.floor(Math.random() * 1000);
    this.screenName = name;
  }
}

export interface PlayerMap {
  [key: string]: Player;
}
