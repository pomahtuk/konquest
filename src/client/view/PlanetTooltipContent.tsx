import React, { ReactElement, useContext } from "react";
import Planet from "../../logic/Planet";
import { css } from "emotion";
import { ThemeContext } from "./themes/ThemeProvider";

interface PlanetTooltipContentProps {
  planet: Planet;
  modifier?: number;
}

const PlanetTooltipContent = ({ planet, modifier = 0 }: PlanetTooltipContentProps): ReactElement => {
  const theme = useContext(ThemeContext);

  const firstTdClass = css`
    padding-right: ${theme.units.medium};
  `;

  const lastTdClass = css`
    text-align: right;
  `;

  return (
    <table
      className={css`
        text-align: left;
      `}
    >
      <tbody>
        <tr>
          <td className={firstTdClass}>Owner:</td>
          <td className={lastTdClass}>{planet.owner ? planet.owner.screenName : "none"}</td>
        </tr>
        <tr>
          <td className={firstTdClass}>Production:</td>
          <td className={lastTdClass}>{planet.production}</td>
        </tr>
        <tr>
          <td className={firstTdClass}>Kill percent:</td>
          <td className={lastTdClass}>{planet.killPercent}</td>
        </tr>
        <tr>
          <td className={firstTdClass}>Ships:</td>
          <td className={lastTdClass}>{planet.ships - modifier}</td>
        </tr>
      </tbody>
    </table>
  );
};

export default PlanetTooltipContent;
