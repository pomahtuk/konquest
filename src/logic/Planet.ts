import Player from "./Player";

export interface PlanetCoordinates {
  x: number;
  y: number;
}

export default class Planet {
  public name: string;
  public owner: number | null = null; // reference co playerId
  public ships: number;
  public shipsDue: number | 0 = 0; // only if there are ships en route
  public production: number; // only happens when planet is taken, next turn
  public killPercent: number;
  // Kill percent is a measure of the effectiveness of the ships produced at that planet.
  // Attack fleets take the kill percentage of their planet of departure,
  // and defense fleets use the kill percentage of the planet they are defending.
  public coordinates: PlanetCoordinates;

  public constructor(name: string, player?: Player | null, coordinates: PlanetCoordinates = { x: 0, y: 0 }) {
    // initialize
    this.name = name;
    this.coordinates = coordinates;

    let production = 10;
    const killPercent = 0.5;
    if (!player) {
      production = 8 + Math.floor(Math.random() * 8);
      // TODO: generate killPercent
      this.ships = 0;
    } else {
      this.owner = player.id;
      this.ships = production;
    }

    this.production = production;
    this.killPercent = killPercent;
  }
}

export interface PlanetMap {
  [key: string]: Planet;
}
