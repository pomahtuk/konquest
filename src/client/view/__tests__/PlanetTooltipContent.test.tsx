import React from "react";
import { render } from "@testing-library/react";

import wrapWithReduxAndStyle from "../testHelpers/wrapWithReduxAndStyle";
import PlanetTooltipContent from "../PlanetTooltipContent";
import Planet from "../../../logic/Planet";
import Player from "../../../logic/Player";

describe("<PlanetTooltipContent />", (): void => {
  it("renders <PlanetTooltipContent /> component for neutral planet", (): void => {
    const testPlanet = new Planet("A");
    testPlanet.ships = 10;
    testPlanet.killPercent = 0.5;
    testPlanet.production = 10;

    const { container } = render(wrapWithReduxAndStyle(<PlanetTooltipContent planet={testPlanet} />));
    expect(container).toMatchSnapshot();
  });

  it("renders <PlanetTooltipContent /> component for player planet", (): void => {
    const testPlayer = new Player("test");
    const testPlanet = new Planet("A", testPlayer);
    testPlanet.ships = 10;
    testPlanet.killPercent = 0.5;
    testPlanet.production = 10;

    const { container } = render(wrapWithReduxAndStyle(<PlanetTooltipContent planet={testPlanet} />));
    expect(container).toMatchSnapshot();
  });

  it("renders <PlanetTooltipContent /> component for player planet and ships modifier", (): void => {
    const testPlayer = new Player("test");
    const testPlanet = new Planet("A", testPlayer);
    testPlanet.ships = 10;
    testPlanet.killPercent = 0.5;
    testPlanet.production = 10;

    const { container } = render(wrapWithReduxAndStyle(<PlanetTooltipContent planet={testPlanet} modifier={10} />));
    expect(container).toMatchSnapshot();
  });
});
