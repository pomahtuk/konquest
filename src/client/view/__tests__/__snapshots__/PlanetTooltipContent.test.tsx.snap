// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlanetTooltipContent /> renders <PlanetTooltipContent /> component for neutral planet 1`] = `
.emotion-0 {
  padding-right: 8px;
}

.emotion-1 {
  text-align: right;
}

.emotion-8 {
  text-align: left;
}

<div>
  <table
    class="emotion-8"
  >
    <tbody>
      <tr>
        <td
          class="emotion-0"
        >
          Owner:
        </td>
        <td
          class="emotion-1"
        >
          none
        </td>
      </tr>
      <tr>
        <td
          class="emotion-0"
        >
          Production:
        </td>
        <td
          class="emotion-1"
        >
          10
        </td>
      </tr>
      <tr>
        <td
          class="emotion-0"
        >
          Kill percent:
        </td>
        <td
          class="emotion-1"
        >
          0.5
        </td>
      </tr>
      <tr>
        <td
          class="emotion-0"
        >
          Ships:
        </td>
        <td
          class="emotion-1"
        >
          10
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`<PlanetTooltipContent /> renders <PlanetTooltipContent /> component for player planet 1`] = `
.emotion-0 {
  padding-right: 8px;
}

.emotion-1 {
  text-align: right;
}

.emotion-8 {
  text-align: left;
}

<div>
  <table
    class="emotion-8"
  >
    <tbody>
      <tr>
        <td
          class="emotion-0"
        >
          Owner:
        </td>
        <td
          class="emotion-1"
        >
          test
        </td>
      </tr>
      <tr>
        <td
          class="emotion-0"
        >
          Production:
        </td>
        <td
          class="emotion-1"
        >
          10
        </td>
      </tr>
      <tr>
        <td
          class="emotion-0"
        >
          Kill percent:
        </td>
        <td
          class="emotion-1"
        >
          0.5
        </td>
      </tr>
      <tr>
        <td
          class="emotion-0"
        >
          Ships:
        </td>
        <td
          class="emotion-1"
        >
          10
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`<PlanetTooltipContent /> renders <PlanetTooltipContent /> component for player planet and ships modifier 1`] = `
.emotion-0 {
  padding-right: 8px;
}

.emotion-1 {
  text-align: right;
}

.emotion-8 {
  text-align: left;
}

<div>
  <table
    class="emotion-8"
  >
    <tbody>
      <tr>
        <td
          class="emotion-0"
        >
          Owner:
        </td>
        <td
          class="emotion-1"
        >
          test
        </td>
      </tr>
      <tr>
        <td
          class="emotion-0"
        >
          Production:
        </td>
        <td
          class="emotion-1"
        >
          10
        </td>
      </tr>
      <tr>
        <td
          class="emotion-0"
        >
          Kill percent:
        </td>
        <td
          class="emotion-1"
        >
          0.5
        </td>
      </tr>
      <tr>
        <td
          class="emotion-0"
        >
          Ships:
        </td>
        <td
          class="emotion-1"
        >
          0
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
